bitwarden-cli: 2024.12.0 -> 2025.1.2

Tobix99 opened the pull request:
Updated Bitwarden CLI to v2025.1.2

Diff: https://github.com/bitwarden/clients/compare/cli-v2024.12.0...cli-v2025.1.2

Changelog: https://github.com/bitwarden/clients/releases/tag/cli-v2025.1.2

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dotlambda reviewed (changes requested):
Please squash your commits and follow https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions.

Tobix99 committed:
bitwarden-cli: 2024.12.0 -> 2025.1.2

Diff: https://github.com/bitwarden/clients/compare/cli-v2024.12.0...cli-v2025.1.2

Changelog: https://github.com/bitwarden/clients/releases/tag/cli-v2025.1.2

dotlambda commented:
The build fails. See recent vsultwarden changes for a possible fix.

Tobix99 commented:
Hi @dotlambda, just FYI I'm currently working on a fix, but am very new to this nix packages (only used `home-manager` until now), so it'll take some time 😀

eclairevoyant commented:
In addition to setting `npmRebuildFlags`, the `noBrokenSymlinks` check is quite strict, hence the broken symlinks in the package are now an error (#380227).

Hence you'll have to somehow correct or remove the following dangling symlinks:

```
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/vault-export-core
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/web-vault
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/vault
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/node
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/platform
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/key-management
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/importer
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/desktop-napi
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/desktop
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/common
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/components
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/browser
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/billing
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/auth
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/admin-console-common
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/admin-console
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/angular
$out/lib/node_modules/@bitwarden/clients/node_modules/@bitwarden/cli
$out/lib/node_modules/@bitwarden/clients/node_modules/.bin/bw
```

Tobix99 commented:
So my problem at the moment is, that the builder is not finding the esbuild package. It tries to download on the fly but fails as it's using the npm offline [flag](https://github.com/NixOS/nixpkgs/blob/6adcf70baf93366e978a44d343fceecac5dcb4a1/pkgs/build-support/node/build-npm-package/hooks/npm-config-hook.sh#L85).

I tried to put it as a input, but then it complained that the version is to new (0.24.1 and it expects 0.23.0).

Also I found, that 2025.1.0 is building fine, but 2025.1.1 is where the build problem starts.

`esbuild` is in the optional dependencies in the `package.json` file, but it's not downloading or finding it in the `node_modules` folder.

eclairevoyant commented:
Use the fix mentioned earlier: set `npmRebuildFlags` as in https://github.com/NixOS/nixpkgs/pull/376765. It works on my machine, I've upgraded locally.

Tobix99 commented:
@eclairevoyant Ok nice will try it out 👍🏻

dotlambda reviewed (dismissed):
The NixOS test fails with `Error: Account does not exist`.

eclairevoyant commented:
> The NixOS test fails with `Error: Account does not exist`.

Not related to this PR, the test itself is broken on f14a17ad45870dbaca282396cfcafb5ef6c99f88

Besides, the cli runs perfectly fine on my machine, even after logging out and back in.

eclairevoyant commented on pkgs/by-name/bi/bitwarden-cli/package.nix:
```diff
@@ -64,6 +69,11 @@ buildNpmPackage rec {
   '';
 
+  postInstall = ''
+    # Clean up broken symlinks.
+    find "$out"/lib/node_modules/ -xtype l -delete
```
This seems like an excessive nit to block on, considering the package _does not build right now_.

Tobix99 replied:
Ok, I'll try to disable the hook 👍 ... but I'm not sure why there are these symlinks in the firstplace ... they all link two directories up, but they are in the `node_modules/@bitwarden` directory so they link to the tools which are in the "root" of the build

dotlambda commented:
> Not related to this PR, the test itself is broken on [f14a17a](https://github.com/NixOS/nixpkgs/commit/f14a17ad45870dbaca282396cfcafb5ef6c99f88); see https://github.com/NixOS/nixpkgs/runs/36596559562

It *is* related. That failure is a different error and flaky

eclairevoyant commented:
> It _is_ related.

Okay, then why does the CLI work perfectly fine outside of the VM? Clearly the test itself is broken and it's unrelated to bw-cli. Did you try running the CLI yourself?



> That failure is a different error

https://logs.ofborg.org/?key=nixos/nixpkgs.379124&attempt_id=278887ba-25ad-48b8-a8a5-8054bfe712cd
```
server # [   33.288443] vaultwarden[1045]: [2025-02-03 16:15:34.857][request][INFO] POST /identity/connect/token
server # [   33.299396] vaultwarden[1045]: [2025-02-03 16:15:34.868][vaultwarden::api::identity][ERROR] Username or password is incorrect. Try again. IP: 2001:db8:1::1. Username: meow@example.com.
server # [   33.307256] vaultwarden[1045]: [2025-02-03 16:15:34.876][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
client # Username or password is incorrect. Try again
(finished: must fail: bw --nointeraction --raw login meow@example.com also_super_secret_ZJWpBKZi668QGt, in 5.66 seconds)
(finished: subtest: can't login to nonexistent account, in 5.66 seconds)
subtest: use the web interface to sign up, log in, and save a password
```

https://logs.ofborg.org/?key=nixos/nixpkgs.378937&attempt_id=28e0439b-52ef-47c9-ae8f-75a7c86a757f

```
server # [   28.510489] vaultwarden[855]: [2025-02-09 21:35:21.967][request][INFO] POST /identity/accounts/prelogin
server # [   28.526156] vaultwarden[855]: [2025-02-09 21:35:21.982][response][INFO] (prelogin) POST /identity/accounts/prelogin => 200 OK
server # [   29.063931] vaultwarden[855]: [2025-02-09 21:35:22.521][request][INFO] POST /identity/connect/token
server # [   29.074763] vaultwarden[855]: [2025-02-09 21:35:22.532][vaultwarden::api::identity][ERROR] Username or password is incorrect. Try again. IP: 2001:db8:1::1. Username: meow@example.com.
server # [   29.084265] vaultwarden[855]: [2025-02-09 21:35:22.542][response][INFO] (login) POST /identity/connect/token => 400 Bad Request
client # Username or password is incorrect. Try again
(finished: must fail: bw --nointeraction --raw login meow@example.com also_super_secret_ZJWpBKZi668QGt, in 6.16 seconds)
(finished: subtest: can't login to nonexistent account, in 6.16 seconds)
subtest: use the web interface to sign up, log in, and save a password
```

How exactly are these errors different?
bw-cli isn't used to create the account, so how is this bw-cli's fault?
The selenium test is clearly broken in some fashion.

dotlambda commented:
> How exactly are these errors different?

Never mind then. I misremembered the error message in the other PR.



> The selenium test is clearly broken in some fashion.

Any idea how to fix it? It works on my machine.

Tobix99 commented:
@dotlambda I rebased onto master to incorporate your fixed to the vaultwarden tests ([Link](https://github.com/NixOS/nixpkgs/pull/380716)), but there is still an error

```
selenium.common.exceptions.ElementClickInterceptedException: Message: Element <button id="newItemDropdown" class="!tw-text-contrast disabled:!tw-text-muted disabled:hover:!tw-text-muted disabled:hover:tw-bg-secondary-300 disabled:hover:tw-border-secondary-300 disabled:hover:tw-no-underline disabled:tw-bg-secondary-300 disabled:tw-border-secondary-300 disabled:tw-cursor-not-allowed focus-visible:tw-ring-2 focus-visible:tw-ring-offset-2 focus-visible:tw-ring-primary-600 focus-visible:tw-z-10 focus:tw-outline-none hover:tw-bg-primary-700 hover:tw-border-primary-700 hover:tw-no-underline tw-bg-primary-600 tw-border-2 tw-border-primary-600 tw-border-solid tw-font-semibold tw-inline-block tw-no-underline tw-px-3 tw-py-1.5 tw-rounded-full tw-text-center tw-transition ng-star-inserted" type="button"> is not clickable at point (1235,44) because another element <p class="tw-mb-2 last:tw-mb-0 !tw-text-sm ng-star-inserted"> obscures it
```

I can try to update the tests further, but you said the tests are working on your machine?

eclairevoyant commented:
> Any idea how to fix it? It works on my machine.

The tests are somehow failing locally for me.
Not sure how to fix it; I went through and checked the element selectors, which are currently correct, and I'm not familiar with debugging selenium.

dotlambda commented:
> I can try to update the tests further, but you said the tests are working on your machine?

No worries. I'll debug the test eventually.
